### PR TITLE
Gas Worst-case Estimation Feature (#286)

### PR DESCRIPTION
**Pull request by agroce**
Will address https://github.com/crytic/echidna/issues/286

### CODE REVIEW

File: lib/Echidna/UI/Report.hs [hunk 1]
```diff
@@ -47,6 +48,11 @@ ppCoverage s | s == mempty = Nothing
              | otherwise   = Just $ "Unique instructions: " ++ show (coveragePoints s)
                                  ++ "\nUnique codehashes: " ++ show (length s)
 
+-- | Pretty-print the gas usage information a 'Campaign' has obtained.
+ppGasInfo :: Map Text (Int, [Tx]) -> Maybe String
+ppGasInfo s | s == mempty = Nothing
+            | otherwise   = Just $ foldr (\t1 t2 -> t1 ++ "\n" ++ t2) "" $ map show $ toList s
```
Comment (arcz) [inline lib/Echidna/UI/Report.hs]: you can also consider `intercalate "\n"` instead of `foldr`

**Comment by agroce** [comment]
Current status is that this works but there are a few major outstanding issues, before it's ready for people.  Then we probably need to think about test generation, to hit the corner cases, since this is often nastier than just finding property violations.

- The call sequences need a shrinker.  They are very hard to read, unless very short and easy to find (the boring cases).  ABI offers some general shrinking code, so I think just adapting the test shrinker in Test.hs to this case will work fairly easily

- The JSON has the sequence, but for some reason I don't get, drops the actual gas values.  Figure out why and include those.

**Comment by agroce** [comment]
- not sure about the function in SolCall being a good map key in multicontract...

**Comment by agroce** [comment]
@ggrieco-tob this is ready for review, I think, assuming we're punting the shrinker for a second PR.  The only thing not checking out ok is the one codeclimate issue that looks to pre-exist the PR.

File: lib/Echidna/Campaign.hs [hunk 2]
```diff
@@ -166,9 +174,38 @@ evalSeq v e = go [] where
     case xs of []     -> pure []
                (y:ys) -> e y >>= \a -> ((y, a) :) <$> go (y:r) ys
 
+-- | Given a call sequence that produces Tx with gas >= g for f, try to randomly generate
+-- a smaller one that achieves at least that gas usage
+shrinkGasSeq :: ( MonadRandom m, MonadReader x m, MonadThrow m
```
Comment (ggrieco-tob) [inline lib/Echidna/Campaign.hs]: This one is unused, right?

Comment (agroce) [reply lib/Echidna/Campaign.hs]: Yes, that's for the future gas-based shrinking
